src/modules/esl-utils/misc/object.ts

```diff
@@ -111,6 +96,75 @@ export const get = (data: any, path: string, defaultValue?: any): any => {
   return typeof result === 'undefined' ? defaultValue : result;
 };
 
+/** Key definition for {@link set} */
+export type PathKeyDef = {
+  /** Key name */
+  key: string | number;
+  /** Key represents collection index */
+  isIndex?: boolean;
+  // /** Key should produce array is not exists */
+  isIndexed?: boolean;
+};
+export type PathKey = PathKeyDef | string | number;
+
+/** @returns PathKeyDef from the PathDef */
+const toKeyDef = (key: PathKey): PathKeyDef => typeof key === 'object' ? key : {key};
+
+/** Parse path to full {@link PathKeyDef} array */
+const parseKeys = (path: string | (number | string | PathKey)[], strict: boolean): PathKeyDef[] => {
+  if (typeof path === 'string' && path && !strict) return parseKeysExt(path);
+  const parts = Array.isArray(path) ? path : (path || '').split('.');
+  return parts.map(toKeyDef);
+};
+
+/** Parse path to the PathKeysDefinition */
+export const parseKeysExt = (path: string): any => {
+  let match;
+  const parts: PathKeyDef[] = [];
+  const matcher = /^([^[.]+)|\.([^[.]*)|\[([^\]]*)]/g;
+  // eslint-disable-next-line no-cond-assign
+  while (match = matcher.exec(path)) {
+    const [, key1, key2, index] = match;
+    if (index !== undefined && (!index || !isNaN(+index))) {
+      parts.push({key: index, isIndex: true});
+    } else {
+      parts.push({key: key1 || key2 || index || ''});
+    }
+  }
+  return parts;
+};
```

Comment:
@exadel-inc/esl-core-team , 
this is the last thing that currently under optimization, everything else is ready to review